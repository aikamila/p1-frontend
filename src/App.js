import logo from './logo.svg';
import './App.css';
import { BrowserRouter as Router, Route, Switch} from 'react-router-dom';
import StartPage from './pages/StartPage';
import PrivateRoute from './utils/PrivateRoute';
import PublicRoute from './utils/PublicRoute';

import AuthPage from './pages/AuthPage';
import HomePage from './pages/HomePage';
import InitialVerificationPage from './pages/InitialVerificationPage';
import UnsuccessfulInitialVerificationPage from './pages/UnsuccessfulInitialVerificationPage';

function App() {
  return (
     <div className="App">

          <PublicRoute path="/" exact>
            <StartPage/>
          </PublicRoute>
          <PublicRoute path="/auth">
            <AuthPage/>
          </PublicRoute>
          <PrivateRoute path="/home"> 
            <HomePage/>
          </PrivateRoute>
            <Switch>
              <Route path="/verification/unsuccessful" component={UnsuccessfulInitialVerificationPage}/>
              <Route path="/verification/:id/:token" component={InitialVerificationPage} />
            </Switch>

     </div>

  );
}

export default App;
