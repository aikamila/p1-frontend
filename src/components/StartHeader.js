import React from 'react'
import { Link } from 'react-router-dom'
import './style/StartHeader.css';
import Logo from './Logo'


const StartHeader = () => {
    return (
        <header className='start-header' role="banner">
            <Logo></Logo>
            <nav className='start-header__navbar' aria-label='main navigation' role="navigation">
            <Link className="start-header__sign-up" role="button" to='/auth/signup'>Sign up</Link>
            <span className='start-header__division'>   |   </span>
            <Link className="start-header__log-in" role="button" to='/auth'>Log in</Link>
            </nav>
        </header>
        
    )
}

export default StartHeader
