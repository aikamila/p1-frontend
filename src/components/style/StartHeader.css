.start-header {
  color: white;
  background-color: black;
  height: 20vh;
  position: -webkit-sticky;
  /* for safari */
  width: 100vw;
  position: relative;
}

.start-header__sign-up,
.start-header__log-in {
  color: white;
}

.start-header__navbar {
  position: relative;
  display: flex;
  flex-direction: row;
  height: 17vh;
}

@media screen and (max-width: 820px) {
  .start-header__navbar {
    justify-content: space-evenly;
    align-items: flex-end;
  }
}

@media screen and (min-width: 821px) {
  .start-header__sign-up {
    flex: 1 1 auto;
  }
  .start-header__log-in {
    flex: 1 1 auto;
  }
  .start-header__division {
    display: none;
  }
}
