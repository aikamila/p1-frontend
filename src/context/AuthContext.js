import { createContext, useState, useEffect, useCallback } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { getItemAndCheckExpiry, setItemwith8MonthExpiry } from "../utils/LocalStorage";
import { calculateTime } from "../utils/Timing";
import { decodeUser } from "../utils/JWTDecode";

const AuthContext = createContext()

export default AuthContext;

export const AuthProvider = ({children}) => {

    const history = useHistory()

    let location = useLocation()

    const [authTokens, setAuthTokens] = useState(null)
    const [userId, setUserId] = useState(null)

    const [loginEmail, setLoginEmail] = useState('')
    const [loginPassword, setLoginPassword] = useState('')

    // const [pwResetEmail, setPwResetEmail] = useState('')
    // const [pwResetPassword, setPwResetPassword] = useState('')
    // const [pwResetPasswordSecond, setPwResetPasswordSecond] = useState('')

    const [loginError, setLoginError] = useState(null) // invalid credentials
 
    const [authServerError, setAuthServerError] = useState(null) // server error while logging in
    const [authSignUpServerError, setAuthSignUpServerError] = useState(null) // server error while signing up
    const [unexpectedLogoutError, setUnexpectedLogoutError] = useState('Something went wrong and you were logged out.')
    const [initialVerificationError, setInitialVerificationError] = useState(null)
    const [initialVerificationServerError, setInitialVerificationServerError] = useState(null)
    const [initialVerificationSuccessful, setInitialVerificationSuccessful] = useState(true)

    const [loginButtonDisabled, setLoginButtonDisabled] = useState(false)

    const BACKEND_DOMAIN = 'https://arcane-spire-03245.herokuapp.com/'
    const TOKENS_OBTAIN_URL = BACKEND_DOMAIN + 'api/user/token/'
    const TOKENS_REFRESH_URL = BACKEND_DOMAIN + 'api/user/token/refresh/'
    const TOKENS_BLACKLIST_URL = BACKEND_DOMAIN + 'api/user/token/blacklist/'
    const INITIAL_EMAIL_VERIFICATION = BACKEND_DOMAIN + 'api/user/token/verification/'

    const login = useCallback(async (e) => {
        e.preventDefault();
        setLoginButtonDisabled(true)
        try{
            const response = await fetch(TOKENS_OBTAIN_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'email': loginEmail, 'password': loginPassword})
            })
            const data = await response.json()
            if (response.status === 401) {
                throw Error('Invalid credentials')
            }else if (response.status === 200){
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setLoginPassword('')
                history.push('/home')
            }else if(loginEmail.length === 0 || loginPassword.length === 0){
                throw Error('Invalid credentials')
            }else{
                throw Error('Something else happened')
            }
        }catch (err){
            switch (err.message){
                case 'Invalid credentials':
                    setLoginError(err.message)
                    setLoginPassword('')
                    break;
                default:
                    setAuthServerError('Upps... unexpected problem with the server occured. Try to refresh the site and log in again :)')
            }
        }
        setLoginButtonDisabled(false)
    }, 
    [TOKENS_OBTAIN_URL, loginEmail, loginPassword, history]
    )

    const logout = useCallback(async () => {
        try{
            await fetch(TOKENS_BLACKLIST_URL,{
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'refresh': authTokens.refresh})
            })}catch(err){
                // doing nothing if the blacklisting of tokens goes wrong on the backend
            }
        const tokens = localStorage.getItem('authTokens')
        if (tokens){
            localStorage.removeItem('authTokens')
        }
        setAuthTokens(null)
        setUserId(null)
        history.push('/')
    }, [TOKENS_BLACKLIST_URL])

    const updateToken = useCallback(async () => {
        try{
            const tokens = getItemAndCheckExpiry('authTokens')
            // we get them from the local storage bc when the user comes back, they don't have the hook set (so we would send null :( )
            const response = await fetch(TOKENS_REFRESH_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'refresh': tokens.refresh})
            })
            const data = await response.json()
            if (response.status !== 200) {
                throw Error('Invalid credentials')
            }else{
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setUserId(decodeUser(data))
            }
        }catch(err){
            logout(); 
            setUnexpectedLogoutError('Something went wrong and you were logged out.')
            setTimeout(() => {
                setUnexpectedLogoutError(null)
            }, 5000)
        }
    }, [TOKENS_REFRESH_URL, logout]
    )

    const verifyEmail = async (id, token) => {
        try{
            const response = await fetch(INITIAL_EMAIL_VERIFICATION, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'id': id, 'token': token})
                })
            if(response.status === 401){
                throw Error('Invalid link')
            }else if(response.status !== 200){
                throw Error("Something else happened")
            }
            const data = await response.json()
            if(response.status===200){
                await logout(); // if there is a user that is logged in, their token should be blacklisted
                setAuthTokens(data);
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setInitialVerificationSuccessful(true)
                setTimeout(() => {
                    setInitialVerificationSuccessful(false)
                }, 5000)
                history.push('/home')
            }
            }catch(err){
                if(err.message === 'Invalid link'){
                    setInitialVerificationError('The link is invalid. If you have already used it, just log into your account.')
                }else(
                    setInitialVerificationServerError('Something went wrong. Please use the link once again.')
                )
                history.push("/verification/unsuccessful")
            }
        
    }

    useEffect(() => {
        const fourMinutes = calculateTime(0, 4, 0)
        let interval = setInterval(() => {
            if (authTokens){
                updateToken()
            }
        }, fourMinutes)
        return () => clearInterval(interval)
    }, [authTokens, updateToken])

    useEffect(() => {
        const updateTokenAtTheBeginning = async () => {
            const tokens = getItemAndCheckExpiry('authTokens')
            if(tokens !== null){
                await updateToken();
            }
        }   
        updateTokenAtTheBeginning()
    }, [updateToken])

    useEffect(() => {
        setLoginPassword('')
        setLoginError(null)
        setAuthSignUpServerError(null)
        setAuthServerError(null)
    }, [location])

    let contextData = {
        login: login,
        loginEmail: loginEmail,
        setLoginEmail: setLoginEmail,
        loginPassword: loginPassword,
        setLoginPassword: setLoginPassword,
        loginError: loginError,
        authTokens: authTokens, 
        logout: logout,
        authServerError: authServerError,
        unexpectedLogoutError: unexpectedLogoutError,
        verifyEmail: verifyEmail,
        initialVerificationError: initialVerificationError,
        initialVerificationServerError: initialVerificationServerError,
        initialVerificationSuccessful: initialVerificationSuccessful,
        setAuthSignUpServerError: setAuthSignUpServerError,
        authSignUpServerError: authSignUpServerError,
        BACKEND_DOMAIN: BACKEND_DOMAIN,
        setLoginError: setLoginError,
        userId: userId,
        updateToken: updateToken,
        loginButtonDisabled: loginButtonDisabled,
    }

    return(
        <AuthContext.Provider value={contextData}>
            {children}
        </AuthContext.Provider>
    )}