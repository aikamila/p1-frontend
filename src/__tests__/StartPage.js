import { render, screen } from '@testing-library/react'
import App from '../App'
import AuthContext, { AuthProvider } from '../context/AuthContext'
import {Router} from 'react-router-dom'
import {createMemoryHistory} from 'history'
import userEvent from '@testing-library/user-event'
import StartHeader from '../components/StartHeader'
import renderer from 'react-test-renderer';

test('sign up link takes the user to the sign up page', () => {
    const history = createMemoryHistory();
    history.push('/')
    render(
        <Router history={history}>
            <StartHeader/>
        </Router>
    )
    userEvent.click(screen.getByRole("button", {name: /sign up/i}))
    expect(history.location.pathname).toBe('/auth/signup')
})

test('log in link takes the user to the sign up page', () => {
    const history = createMemoryHistory();
    history.push('/')
    render(
        <Router history={history}>
            <StartHeader/>
        </Router>
    )
    userEvent.click(screen.getByRole("button", {name: /log in/i}))
    expect(history.location.pathname).toBe('/auth')
})


test('the start page is rendered correctly', () => {
    const history = createMemoryHistory();
    history.push('/')
    const page = renderer
    .create(
        <Router history={history}>
            <AuthContext.Provider value={{authTokens: null}}>
                <App/>
            </AuthContext.Provider>
        </Router>
    )
    .toJSON()
    expect(page).toMatchSnapshot()
})
