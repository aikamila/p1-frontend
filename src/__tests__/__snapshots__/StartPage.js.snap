// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the start page is rendered correctly 1`] = `
<div
  className="App"
>
  <header
    className="start-header"
    role="banner"
  >
    <img
      alt="shangify brand logo"
      className="logo__img1"
      src="shangify_logo.png"
    />
    <img
      alt=""
      className="logo__img2"
      src="shangify_brand_name.png"
    />
    <nav
      aria-label="main navigation"
      className="start-header__navbar"
      role="navigation"
    >
      <a
        className="start-header__sign-up"
        href="/auth/signup"
        onClick={[Function]}
        role="button"
      >
        Sign up
      </a>
      <span
        className="start-header__division"
      >
           |   
      </span>
      <a
        className="start-header__log-in"
        href="/auth"
        onClick={[Function]}
        role="button"
      >
        Log in
      </a>
    </nav>
  </header>
  <main
    className="start-page"
  >
    <div
      className="start-page__beginning"
    >
      <video
        autoPlay={true}
        height="480"
        loop={true}
        muted={true}
        width="640"
      >
        <source
          src="i_love_this_vid.mp4"
        />
        This video is incompatible with your browser.
      </video>
      <h1>
        You always wanted to change the world.
      </h1>
      <h2>
        We made it possible for YOU.
      </h2>
      <button
        onClick={[Function]}
      >
        See how it works
      </button>
    </div>
    <div
      className="start-page__sections"
    >
      <div
        className="animate__animated"
        style={
          Object {
            "animationDuration": "2s",
            "opacity": 0,
          }
        }
      >
        <section
          className="start-page__section1"
        >
          <h3>
            What is this app really about?
          </h3>
          <p>
            We believe that all people are capable of great things. Most of them are too scared or too doubtful to realize their plans. It doesn't matter whether you would like to start a charity organization or maybe set up your own company - we know that we are all sstronger together. We give you an amazing opportunity to put your dreams into action with 
            <nobr>
              like-minded
            </nobr>
            ,
            <nobr>
              goal-oriented
            </nobr>
             people.
          </p>
        </section>
      </div>
      <div
        className="animate__animated"
        style={
          Object {
            "animationDuration": "2s",
            "opacity": 0,
          }
        }
      >
        <section
          className="start-page__section2"
        >
          <h3>
            How can I start?
          </h3>
          <p>
            Sign up, verify your email and start - it's as simple as that! You can share your ideas publically or take a look at other people's posts - you can either comment or simply look for inspiration. Unleash your potential!
          </p>
        </section>
      </div>
    </div>
  </main>
</div>
`;
